api/mutations/channel/utils.js

```diff
@@ -0,0 +1,38 @@
+// @flow
+import createLoaders from '../../loaders/';
+const loaders = createLoaders();
+
+export const userCanManageChannel = async (
+  userId: string,
+  channelId: string
+): Promise<boolean> => {
+  if (!userId || !channelId) return false;
+
+  const channel = await loaders.channel.load(channelId);
+
+  if (!channel || channel.deletedAt) return false;
+  const [communityPermissions, channelPermissions] = await Promise.all([
+    loaders.userPermissionsInCommunity.load([userId, channel.communityId]),
+    loaders.userPermissionsInChannel.load([userId, channelId]),
+  ]);
+
+  if (!communityPermissions) return false;
+  if (communityPermissions.isOwner || communityPermissions.isModerator)
+    return true;
+  if (!channelPermissions) return false;
+  if (channelPermissions.isOwner || channelPermissions.isModerator) return true;
+  return false;
+};
+
+export const userCanCreateChannel = async (
```

Comment:
Rather than tying this to a specific mutation why don't we call this `userOwnsCommunity`? (which is really what it checks) That way we can reuse it for all the community mutations too.

I don't think it makes sense to externalise all permissions handling outside of the resolvers, I think instead these should be used as helper methods inside of the resolvers themselves.

Reply:
To elaborate a bit here since this might not be immediately obvious:

```JS
// mutations/channel/createChannel.js
const resolver = (/* ... */) => {
  if (!(await userCanCreateChannel(input.communityId))) { /* ... */ }
}

// vs.

const resolver = (/* ... */) => {
  if (!(await userOwnsCommunity(input.communityId))) { /* ... */ }
}
```

The first one externalises permissions handling and requires you to go look up a different file to figure out when you're allowed or not to create a channel. This indirection is really unnecessary, as the `userCanCreateChannel` method won't ever be used anywhere.

The second one looks similar on the surface, but is very different in that it only adds some nice helper sugar on top of our permissions system. Importantly, it doesn't externalise the permissions handling at all, it's all still in the resolver and doesn't require me to look up a different file to figure out when I'm allowed to create a channel.

Reply:
@mxstbr FWIW, I just think directives are the wrong solution to the problem of "I have too much repetitious logic".

Using Graphene at the moment, so I don't have a Node example, but this is how we're handling fields that require auth:

```python
@anonymous_return(list)
def resolve_foo(root, info, **args):
    return whatevs
```

The decorator accepts a direct value or a callable, and returns that if you attempt to resolve the field without an authenticated user.

This has nice advantages, because it's super-easy return things like empty arrays, nulls, zero values or whatever, rather than just a yes/no check of whether anonymous users are allowed to do something. I also think it's wrong for a GraphQL server to *ever* throw when executing a query unless something is actually broken. All user-facing error outcomes should be modelled into the schema.

I use this on both regular fields and mutations. In the case of mutations, I always use unions to represent success/failure payloads, so I tell anonymous_return to immediately return the failure payload:

```python
@anonymous_return(CreateFooFailure('You must be logged in to create a foo'))
def resolve_create_foo(root, info, **args):
    return whatevs
```

Reply:
@AndrewIngram Appreciate the note! This specific comment is not about directives, these are just helper utils we use in the resolvers themselves. 

```JS
const resolver = (/* ... */) => {
  if (!(await userOwnsCommunity(input.communityId))) { /* ... */ }
}
```

The only directive that was added here is `isAuthed`, which only checks to make sure that a user is authenticated before running mutations—we won't move the permissions stuff to directives I'm pretty sure. :blush: 

> I also think it's wrong for a GraphQL server to ever throw when executing a query unless something is actually broken. All user-facing error outcomes should be modelled into the schema.

This is interesting, because we definitely don't do that right now. We have a `UserError` class which we throw when we want the error message to be exposed to the user, and any other error we send a generic `Internal server error` and log it to Sentry. What's the benefit of modelling that in the schema itself?

Reply:
If you want to show a message to a user, you already have a mechanism for defining a contract between client and server -- it's your GraphQL schema. By moving user-facing data outside of the schema and into the errors array, you're back into the world of ad-hoc data structures and implicit conventions.

When you have a powerful expressive type system (GraphQL's is no Elm or Haskell, but it's decent), errors-as-data is beautifully elegant way of working, and pairs really nicely with functional programming. Errors that interrupt program flow (throwing exceptions that only get caught by a top-level) aren't as nice to work with, so should really be for the "something has unexpectedly gone wrong" scenario -- the equivalent of a 500 error.

You've seen all the Elm talks about making impossible states impossible, and allowing your types to force you to handle empty/loading/error states. This is the power of a comprehensive type system, it creates more work, but results in better software.

Reply:
Do you have a specific example of what that would look like in the schema? Do you just have a generic Error type?

```GraphQL
type Thread {
  author: User | Error
}
```

and then from the frontend you do something like

```JS
if (props.data.error) {
  // render error state
}
```

if I understand that correctly?

Reply:
I have a mixture of generic and non-generic errors, let's look at a deliberately complex example.

Say I have a mutation for a form, let's call it `addUser`, and you need to be both authenticated and have certain permissions, the client would handle anonymous users differently from users without permission (show login form versus show an access denied message). I might choose to model it this way:

```graphql
type LoginRequired {
  message: String!
}

type InsufficientPermissions {
  message: String!
}

type FieldError {
  fieldName: String!
  messages: [String!]! 
}

type FormValidationFailure {
   fieldErrors: [FieldErrors!]!
   nonFieldErrors: [String!]!
}

type AddUserSuccess {
   user: User!
   ... other fields
}

union AddUserPayload = LoginRequired | InsufficientPermissions | FormValidationFailure | AddUserSuccess

type mutation {
  ...
  addUser(username: String!, email: String!, password: String!): AddUserPayload!
  ...
}
```

It might be a bit overwhelming, but hopefully you can see that the `AddUserPayload` union is capable of modelling a range of scenarios from super-generic, to super-specific. Some of the types will be shared with other mutations and query field, some might be specific to the mutation at hand. Similar to how your React components will vary widely in the degree to which they get reused, your error modelling can work in the same way.